doc/designs/lms.md

```diff
@@ -0,0 +1,243 @@
+security of its underlying hash functions, and approved secure hash functions
+are believed to be quantum resistant.
+Stateful HBS schemes are not suitable for general use because they require
+careful state management that is often difficult to ensure.
+
+Overview
+---------
+
+A signature is used to validate the authenticity of a message by associating a
+secret private key with a shared public key. Leighton-Micali One Time Signatures
+(LM-OTS) require a unique private key to sign any given message. This means that
+a LMS private key must contain state information so that it changes the LM-OTS
+private key each time a message is signed. The system is not secure if the same
+private key is used more than once.
+
+The LMS scheme uses a Merkle hash tree, where the leaves of the tree are LM-OTS
+public keys, and each non leaf node is a hash of its 2 children. The root node
+is the shared public key of the tree. For a h height tree there are 2^h OTS
+public keys.
+
+A LM-OTS signature uses Hash functions that use a OTS private key and message.
+A LMS signature consists of a leaf index q combined with a LM-OTS signature and
+a path of adjacent hashes going from the leaf node up to the root. The signature
+can then be verified by comparing the supplied public key to a calculated
+public key.
+
+Since all leaf nodes are required to obtain a public key for a single
+LMS tree, a very large tree can take a long time to generate.
+LMS has a multi-tree variant called the Hierarchical Signature System (HSS)
+which allows smaller subtrees to be built as required.
+In HSS the public key for the first LMS tree is the public key of the HSS system.
+Each LMS private key signs the next levels LMS public key, and a bottom level
+LMS tree signs the actual message using a leaf LM-OTS private key.
+
+Refer to [1] & [4] for further information.
+
+What is supported
+-----------------
+
+- HSS public keys and HSS signatures. LMS and HSS public keys may share the
+same object, there is only one additional field L for HSS public keys.
+- HSS signature verification
+- Support for modes and digests specified in [2].
+
+What is not supported
+---------------------
+
+- LMS key pair generation, signature generation and private key support.
+  Due to the problematic nature of managing the stateful OTS private keys,
+  [2] requires that key and signature generation be performed in
+  hardware cryptographic modules that do not allow secret keying material to be
+  exported, even in encrypted form.
+- LMS and LM_OTS signatures.
+  The code will be written in a way that would allow LMS signatures to be added
+  if required, but a HSS signature with only a single level tree should be
+  sufficient for this use case. A separate LMS signature dispatch table could be
+  added if this is ever required.
+- CMS and X509 support - this may be added at a later date.
+- Implementing Extended Merkle signature scheme (XMSS) is a separate effort.
```

Comment:
Yup. HSS with a single tree is basically LMS. No one uses bare LMS, so that's fine. Definitely good not so support signatures and key gen. When implementing XMSS, it's possible to use some common code, but LMS and XMSS is just different enough that it's probably best to have separate implementations (said by someone who has made the combined implementation as a testbed).